content/en/docs/ops/common-problems/network-issues/index.md

```diff
@@ -233,6 +233,115 @@ server {
 }
 {{< /text >}}
 
+## 404 or 503 error while accessing headless services
+
+Assume Istio is installed with the following configuration:
+
+- `mTLS mode` set to `STRICT` within the mesh
+- `meshConfig.outboundTrafficPolicy.mode` set to `ALLOW_ANY`
+
+Consider `nginx` is deployed as a `StatefulSet` in the default namespace and a corresponding `Headless Service` is defined as shown below:
+
+{{< text yaml >}}
+apiVersion: v1
+kind: Service
+metadata:
+  name: nginx
+  labels:
+    app: nginx
+spec:
+  ports:
+  - port: 80
+    name: web      # Port name
+  clusterIP: None  # Creates a Headless Service
+  selector:
+    app: nginx
+---
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: web
+spec:
+  selector:
+    matchLabels:
+      app: nginx
+  serviceName: "nginx"
+  replicas: 3
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - name: nginx
+        image: k8s.gcr.io/nginx-slim:0.8
+        ports:
+        - containerPort: 80
+          name: web
+{{< /text >}}
+
+Let us assume we have a [sleep]({{< github_tree >}}/samples/sleep) pod `Deployment` as well in the default namespace.
+
+Now, let us consider 3 possible scenarios based on the port `name` defined in the headless service:
+
+1. Port `name` set to `web` (as shown in the example above) or any other custom name
+
+    - Request without Host header:
+
+        When `nginx` is accessed without the Host header, the sidecar proxy on the client-side fails to find the route entry to `nginx` and fails with `HTTP 404 NR`.
+
+        {{< text bash >}}
+        $ kubectl exec -it sleep-557747455f-blkgt -c sleep -- sh
+        / $ curl 10.1.1.171 -s -o /dev/null -w "%{http_code}"
+          404/
+        {{< /text >}}
+
+        Here, `10.1.1.171` is the Pod IP of one of the replicas of `nginx` and the service is accessed on `containerPort` 80.
+    - Request with Host header:
+
+        When `nginx` is accessed with the correct Host header, it successfully returns `HTTP 200 OK`.
+
+        {{< text bash >}}
+        $ kubectl exec -it sleep-557747455f-blkgt -c sleep -- sh
+        / $ curl -H "Host: nginx.default" 10.1.1.171 -s -o /dev/null -w "%{http_code}"
```

Comment:
A more common way to handle this would probably be to do `curl nginx-0.nginx.default`. Both work, but one is standard way to access specific instance of a headless service via hostname and the other is more or less a workaround to appease Istio.

Reply:
Good to know that _curl nginx-0.nginx.default_ works fine, the issue is only when pod IP is used. Based on the comment here https://github.com/istio/istio/issues/34679#issuecomment-898561039, I understand that when the port name is explicitly set to "http", the problem described in my 2nd section will **still** exist. However, if using pod IP is not the right way to access a headless service, should I rephrase or remove the 2nd section (with port name set to http)?